torch_xla/distributed/xrt_init.py

```diff
@@ -0,0 +1,241 @@
+    os.environ[xenv.ORDINAL] = str(rank)
+    os.environ[xenv.LOCAL_ORDINAL] = str(local_rank)
+    os.environ[xenv.LOCAL_WORKER] = 'localservice:' + str(rank)
+
+    os.environ[xenv.MP_DEVICE] = f'GPU:{rank}'
+    gpus_to_use = os.environ.get('CUDA_VISIBLE_DEVICES')
+    if gpus_to_use is not None:
+        # If gpu devices are set by a scheduling entity (eg. SLURM) we index into
+        # comma separated string containing numbered gpu devies
+        gpus_to_use_list = gpus_to_use.split(',')
+        os.environ['CUDA_VISIBLE_DEVICES'] = gpus_to_use_list[local_rank]
+    else:
+        # If no explicit visible devices are provided, local_rank is used to identify
+        # the gpu used by this process
+        os.environ['CUDA_VISIBLE_DEVICES'] = str(local_rank)
+
+
+def _init_xrt_context(master_addr=None, master_port=None, store=None):
+    '''
+    Initializes the XLA device depending on the kind of the device. Or is a no-op if init_xrt_context
+    has already been called.
+
+    Keyword Arguments:
+    master_addr -- int(SomeAddrNumber) (default tries to get addr from
+                    environment variable, used to setup TCPStore)
+                    Not needed if store is provided.
+    master_port -- int(SomePortNumber) (default tries to get port from
+                    environment variable, used to setup TCPStore)
+                    Not needed if store is provided.
+    store       -- A TCPstore object (default None) -- If None a TCPStore object
+                    will be setup for you
+    '''
+    global _INIT_XRT_ALREADY_CALLED
+
+    if _INIT_XRT_ALREADY_CALLED:
+        return
+
+    # Call this in the actual test case, to work with torch/xla workers
+    rank = int(os.environ['RANK'])
+    local_rank = int(os.environ['LOCAL_RANK'])
+    world_size = int(os.environ['WORLD_SIZE'])
+    group_rank = int(os.environ['GROUP_RANK'])
+    local_world_size = int(os.environ['LOCAL_WORLD_SIZE'])
+
+    if master_addr is None:
+        master_addr = os.environ['MASTER_ADDR']
+
+    if master_port is None:
+        master_port = os.environ['MASTER_PORT']
+
+    try:
+        _, dev_kind = _get_devices_per_worker()
+    except RuntimeError as e:
+        # Defaulting to TPU if no env variable is set
+        dev_kind = 'TPU'
+    os.environ.pop(xenv.TPU_CONFIG, None)
+    os.environ.pop(xenv.TPU_NUM_DEVICES, None)
+    os.environ.pop(xenv.GPU_NUM_DEVICES, None)
+
+    # This is required if we want to dynamically grab free ports.
+    # Useful in shared settings when we cannot predetermine what ports are taken.
+    is_server = True if rank is '0' else False
+    global _TCP_STORE
+    if store is None:
+        assert master_addr is not None
+        assert master_port is not None
+        _TCP_STORE = dist.TCPStore(master_addr, int(master_port), world_size,
+                                   is_server)
+    else:
+        _TCP_STORE = store
+
+    node_list = None
+
+    if dev_kind == 'TPU':
```

Comment:
Why are the two `if dev_kind == 'TPU'` blocks separated here? Does part need to run before `_set_mesh_config` and half need to run after?